### ISSUE TITLE
Infinite recursion in `Match -> participant -> champion`

### ISSUE THREAD

**Issue by pfmoore** [meraki-analytics/cassiopeia#154]
This may be somehow related to my setup, but I'm not sure how...

Simple example (EUW):

```
>>> from cassiopeia import apply_settings
>>> apply_settings('settings.json')
>>> from cassiopeia.core import Match
>>> # Random match ID from ones I have cached
>>> m = Match(id=3374776468)
>>> # Check I get the cached data
>>> m.season
<Season.season_7: 'SEASON2017'>
>>> m.participants[0].champion
... huge traceback, mostly lost to the limited scroll buffer in my console ...
...
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 113, in _transform
    data = transformer.transform(target_type, data, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 105, in transform
    return transformer.transform(target_type, value, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 50, in wrapper
    return call(self, value, context=context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\transformers\staticdata.py", line 202, in champion_list_data_to_core
    return Champions([self.champion_data_to_core(c) for c in value], region=value.region, version=value.version, locale=value.locale, included_data=value.included_data)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\core\common.py", line 41, in default_region_wrapper
    return method(self, *args, **kwargs)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\core\common.py", line 175, in __call__
    return pipeline.get(cls, query=query)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 433, in get
    return handler.get(query, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 190, in get
    sink.put(result, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 138, in put
    item = self._transform(data=item, context=context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 113, in _transform
    data = transformer.transform(target_type, data, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 105, in transform
    return transformer.transform(target_type, value, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 50, in wrapper
    return call(self, value, context=context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\transformers\staticdata.py", line 202, in champion_list_data_to_core
    return Champions([self.champion_data_to_core(c) for c in value], region=value.region, version=value.version, locale=value.locale, included_data=value.included_data)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\core\common.py", line 41, in default_region_wrapper
    return method(self, *args, **kwargs)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\core\common.py", line 175, in __call__
    return pipeline.get(cls, query=query)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 433, in get
    return handler.get(query, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 190, in get
    sink.put(result, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 138, in put
    item = self._transform(data=item, context=context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\pipelines.py", line 113, in _transform
    data = transformer.transform(target_type, data, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 105, in transform
    return transformer.transform(target_type, value, context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\datapipelines\transformers.py", line 50, in wrapper
    return call(self, value, context=context)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\cassiopeia\transformers\staticdata.py", line 51, in champion_list_dto_to_data
    data = deepcopy(value)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 306, in _reconstruct
    value = deepcopy(value, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 306, in _reconstruct
    value = deepcopy(value, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 215, in _deepcopy_list
    append(deepcopy(a, memo))
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\copy.py", line 146, in deepcopy
    cls = type(x)
RecursionError: maximum recursion depth exceeded while calling a Python object
```

I've had this with a couple of different match IDs, so I don't *think* it's related to the match. The only specific thing in mt settings is that I have a local file cache enabled. The match is in the cache, but all I've pulled is the season data (i.e. the base match object). The first time I did this I triggered a call to Data Dragon for the champion list. That's now cached. Otherwise, I don't know of anything I've done weird. I'm using the latest cass from github.

Let me know if you need any further information.

**Comment by pfmoore** [comment]
Found the following error in the process of trying to simplify the traceback:

```
>>> m.participants[0].champion
Traceback (most recent call last):
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\site-packages\merakicommons\cache.py", line 16, in wrapper
    return values[self]
  File "C:\Users\UK03306\.virtualenvs\d1355cddb7f7f27\lib\weakref.py", line 394, in __getitem__
    return self.data[ref(key)]
KeyError: <weakref at 0x00000000073917C8; to 'Participant' at 0x0000000005223518>
```

**Comment by jjmaldonis** [comment]
We've run into similar recursion errors quite a bit in the past, and have talked a lot about the underlying cause. We finally decided to just simplify some functionality in Cass to make these types of issues go away. I updated my code locally a few days ago and have been using it quite a bit to test it, and I believe it's all working. I just pushed a minute ago.

That commit comes with a boatload of other changes -- but the two you will need to worry about are the `Patch` and `Queue` changes. `Patch` now requires a region in it's constructors, so you'll have to add that parameter if you aren't using `cass.set_default_region`. The enum `Queue` names have also changed to be consistent with Riot's renaming, and since that broke some of our previous names of queues, we just renamed them all (they are so much easier to understand now).

All of the other changes should be invisible. In addition, you can remove the `UnloadedGhostStore()` from your data pipeline, because we now insert it automatically if it isn't there.

Hopefully these backend/invisible changes clean up all of the recursion (and similar) errors that I've seen too much of. We are continuing to better define the lines between all the different components in Cass, and that's what these backend changes have been done.

So tldr; repull from github and these errors should go away. You also might have to update your usage of `Queue` and `Patch`.

**Comment by pfmoore** [comment]
Yes, updated and it's all working fine now. Many thanks.

And in case I hadn't mentioned, the data pipeline persistent cache is an absolutely awesome feature. Yesterday I ran a huge job to download a whole batch of match data, and I can now use it with no delays, no API calls, and no code changes needed. Beautiful :-) One thing I might do is see if I can modify the diskstore backend to create a SQLite database version - the simplekv package has a SQLAlchemy backend, so it should be relatively easy. And I feel that Windows might prefer me not to have directories with 50k of files in them... ;-) If I get it working, I'll send a PR in.
